crates/wasi-common/src/sys/windows/host_impl.rs

```diff
@@ -1,36 +1,45 @@
 //! WASI host types specific to Windows host.
-use crate::{wasi, Error, Result};
+use crate::{error::FromRawOsError, Error, Result};
 use std::ffi::OsStr;
 use std::os::windows::ffi::OsStrExt;
+use winx::winerror::WinError;
 
-pub(crate) fn errno_from_win(error: winx::winerror::WinError) -> wasi::__wasi_errno_t {
-    // TODO: implement error mapping between Windows and WASI
-    use winx::winerror::WinError::*;
-    match error {
-        ERROR_SUCCESS => wasi::__WASI_ERRNO_SUCCESS,
-        ERROR_BAD_ENVIRONMENT => wasi::__WASI_ERRNO_2BIG,
-        ERROR_FILE_NOT_FOUND => wasi::__WASI_ERRNO_NOENT,
-        ERROR_PATH_NOT_FOUND => wasi::__WASI_ERRNO_NOENT,
-        ERROR_TOO_MANY_OPEN_FILES => wasi::__WASI_ERRNO_NFILE,
-        ERROR_ACCESS_DENIED => wasi::__WASI_ERRNO_ACCES,
-        ERROR_SHARING_VIOLATION => wasi::__WASI_ERRNO_ACCES,
-        ERROR_PRIVILEGE_NOT_HELD => wasi::__WASI_ERRNO_NOTCAPABLE, // TODO is this the correct mapping?
-        ERROR_INVALID_HANDLE => wasi::__WASI_ERRNO_BADF,
-        ERROR_INVALID_NAME => wasi::__WASI_ERRNO_NOENT,
-        ERROR_NOT_ENOUGH_MEMORY => wasi::__WASI_ERRNO_NOMEM,
-        ERROR_OUTOFMEMORY => wasi::__WASI_ERRNO_NOMEM,
-        ERROR_DIR_NOT_EMPTY => wasi::__WASI_ERRNO_NOTEMPTY,
-        ERROR_NOT_READY => wasi::__WASI_ERRNO_BUSY,
-        ERROR_BUSY => wasi::__WASI_ERRNO_BUSY,
-        ERROR_NOT_SUPPORTED => wasi::__WASI_ERRNO_NOTSUP,
-        ERROR_FILE_EXISTS => wasi::__WASI_ERRNO_EXIST,
-        ERROR_BROKEN_PIPE => wasi::__WASI_ERRNO_PIPE,
-        ERROR_BUFFER_OVERFLOW => wasi::__WASI_ERRNO_NAMETOOLONG,
-        ERROR_NOT_A_REPARSE_POINT => wasi::__WASI_ERRNO_INVAL,
-        ERROR_NEGATIVE_SEEK => wasi::__WASI_ERRNO_INVAL,
-        ERROR_DIRECTORY => wasi::__WASI_ERRNO_NOTDIR,
-        ERROR_ALREADY_EXISTS => wasi::__WASI_ERRNO_EXIST,
-        _ => wasi::__WASI_ERRNO_NOTSUP,
+impl FromRawOsError for Error {
+    fn from_raw_os_error(code: i32) -> Self {
+        Self::from(WinError::from_u32(code as u32))
+    }
+}
+
+impl From<WinError> for Error {
+    fn from(err: WinError) -> Self {
+        // TODO: implement error mapping between Windows and WASI
```

Comment:
Is this TODO still to be done?

Reply:
I left it there as a reminder since we don't cover all Windows error codes. Then again, we've got the vast majority of WASI syscalls mapped into Windows, so I'm happy to remove it. The point in favour of leaving it in would be that we still may encounter unmapped Windows-to-WASI error codes when adding new WASI tests in, etc. But I'm happy to go either way with this one :-)